File: FileManager.vb
Imports System.IO
Imports System.IO.Compression
''' <summary>A Class to maintain a file abstract of the Grooper Tree Under the git project in order to leverage GIT commands against the Grooper Nodes.
''' </summary><remarks>This is because Grooper Nodes aren't files and need to be able to detect differences as files and compare differences as files</remarks>
Public Class FileManager

    ''' <summary>
    ''' Syncs the active node settings of all chidlren in the project and marks there appropriate GitStatus
    ''' </summary>
    ''' <param name="grooperNode"></param>
    Public Shared Sub SyncProject(grooperNode As GrooperNode)
        If grooperNode.TypeDisplayName = "GitProject" Then

            Dim gitProject = DirectCast(grooperNode, GitProject)
            If gitProject.localRepository Is Nothing Or "" Then Exit Sub
            Dim gitConsole As New GitObject(gitProject.localRepository)
            For Each node As GrooperNode In gitProject.AllChildren
                If Not IO.File.Exists("") Then
                    node.SetValue("GitStatus", "New")
                    Continue For
                End If
                FileManager.NodeToFile(node)
            Next
        Else
        End If
        'TODO:  Write execution code that will be performed on the configured interval
    End Sub
    Public Shared Sub NodeToFile(Obj As Object)
        If TypeOf (Obj) Is GitProject Then
            Dim node As GitProject = Obj
            File.WriteAllLines(node.localRepository & "\" & node.Id.ToString & ".json", node.PropertiesJson.Split(vbLf))
        Else
            Dim node As GrooperNode = DirectCast(Obj, GrooperNode)
            Dim parentNode As GitProject = node.ParentProject
            File.WriteAllLines(parentNode.localRepository & "\" & node.Id.ToString & ".json", node.PropertiesJson.Split(vbLf))
        End If
    End Sub
End Class

File: Classes/FileManager.cs
using System;
using System.IO;
using System.Collections.Generic;
using Grooper;
namespace GrooperGit
{
    /// <summary>
    /// A Class to maintain a file abstract of the Grooper Tree Under the git project in order to leverage GIT commands against the Grooper Nodes.
    /// </summary>
    /// <remarks>
    /// This is because Grooper Nodes aren't files and need to be able to detect differences as files and compare differences as files.
    /// </remarks>
    [Obsolete("FileManager is depricated, factory replaced for mutator NodeAsfile")]
    public class FileManager
    {
        /// <summary>
        /// Syncs the active node settings of all children in the project and marks their appropriate GitStatus.
        /// </summary>
        /// <param name="grooperNode">The GrooperNode to be synced.</param>
        public static void SyncProject(GrooperNode grooperNode)
        {
            if (grooperNode.TypeDisplayName == "GitProject")
            {
                GitProject gitProject = (GitProject)grooperNode;

                // if (string.IsNullOrEmpty(gitProject?.LocalRepository)) return;

                foreach (GrooperNode node in gitProject.AllChildren)
                {
                    if (!File.Exists(""))
                    {
                        node.SetValue("Git", "New");
                        continue;
                    }
                    NodeToFile(node);
                }
            }
        }
        ///<summary>Checks the state of the node </summary>
        ///<remarks>will create the necessary files in the git repository if needed, and 
        ///also set the node value to the appropriate git status</remarks>
        ///<param name="SyncChildrenToo">optionally sync all children of node</param>
        ///<param name="node">The node to be synced with the filesystem></param>
        public static void Sync(GrooperNode node, bool SyncChildrenToo = false)
        {
            List<GrooperNode> nodes = new List<GrooperNode>();
            nodes.Add(node);
            if (SyncChildrenToo == true)
            {
                nodes.AddRange(node.AllChildren);
            }
            foreach(GrooperNode n in nodes)
            {
                NodeToFile(n);
            }
        }
        
        public static void NodeToFile(object obj)
        {
            if (obj is GitProject)
            {
                var node = obj as GitProject;
                // File.WriteAllLines(Path.Combine(node.LocalRepository, node.Id + ".json"), node.PropertiesJson.Split('\n'));
            }
            else
            {
                var node = obj as GrooperNode;
                var parentNode = node.ParentProject as GitProject;
                // File.WriteAllLines(Path.Combine(parentNode.LocalRepository, node.Id + ".json"), node.PropertiesJson.Split('\n'));
            }
        }
    }
}
